src/main/java/org/testng/internal/MethodHelper.java

```diff
@@ -237,8 +252,23 @@ public int compare(Node<ITestNGMethod> o1, Node<ITestNGMethod> o2) {
       String[] methodsDependedUpon = m.getMethodsDependedUpon();
       String[] groupsDependedUpon = m.getGroupsDependedUpon();
       if (methodsDependedUpon.length > 0) {
-        ITestNGMethod[] methodsNamed =
-          MethodHelper.findDependedUponMethods(m, methods);
+        ITestNGMethod[] methodsNamed = null;
+        // Method has instance
+        if (m.getInstance() != null) {
+          // Get other methods with the same instance
+          List<ITestNGMethod> instanceMethods = testInstances.get(m.getInstance());
+          try {
+            // Search for other methods that depends upon with the same instance
+            methodsNamed = MethodHelper.findDependedUponMethods(m, instanceMethods);
+          } catch (TestNGException e) {
+            //Maybe this method has a dependency on a method that resides in a different instance.
+            //Lets try searching for all methods now
+            methodsNamed = MethodHelper.findDependedUponMethods(m, methods);
+          }
+        } else {
+          // Search all methods
+          methodsNamed = MethodHelper.findDependedUponMethods(m, methods);

```

Comment:
I was trying to find a way to avoid the duplication of this line with line 266 but because of the exception, I don't really see an elegant refactoring.